content/doc/book/pipeline/jenkinsfile.adoc

```diff
@@ -889,6 +854,97 @@ within a single `withCredentials( ... ) { ... }` step by doing the following:
   `withCredentials( ... ) { ... }` step snippet.
 
 
+=== String interpolation
+
+Jenkins Pipeline uses rules identical to link:http://groovy-lang.org[Groovy] for
+string interpolation. Groovy's String interpolation support can be confusing to
+many newcomers to the language. While Groovy supports declaring a string with
+either single quotes, or double quotes, for example:
+
+[source,groovy]
+----
+def singlyQuoted = 'Hello'
+def doublyQuoted = "World"
+----
+
+Only the latter string will support the dollar-sign (`$`) based string
+interpolation, for example:
+
+[source,groovy]
+----
+def username = 'Jenkins'
+echo 'Hello Mr. ${username}'
+echo "I said, Hello Mr. ${username}"
+----
+
+Would result in:
+
+[source]
+----
+Hello Mr. ${username}
+I said, Hello Mr. Jenkins
+----
+
+Understanding how to use string interpolation is vital for using some of
+Pipeline's more advanced features.
+
+[WARNING]
+.Interpolation of sensitive environment variables
+====
+
+Groovy string interpolation should [red]*never* be used with sensitive environment variables (i.e. credentials, see: <<Handling credentials>>).
+This is because the sensitive environment variable will be interpolated at Groovy evaluation, and the environment variable's value could be made available earlier than intended, resulting in the sensitive data leaking.
```

Comment:
```suggestion
This is because the sensitive environment variable will be interpolated during Groovy evaluation, and the environment variable's value could be made available earlier than intended, resulting in sensitive data leaking in various contexts.
```

Reply:
resovled in de21b90b62357588df4ac3af80df66263a0c6e7a